### ISSUE TITLE
Suspicion that `@reject_recursive_repeats` is not thread-safe

### ISSUE THREAD

**Issue by reverendus** [ethereum/web3.py#443]
* Version: 3.16.2
* Python: 3.6
* OS: osx

### What was wrong?

When executing a few contract calls concurrently, I kept getting errors like:

```
File "/usr/local/lib/python3.6/site-packages/web3/contract.py", line 870, in call_contract_function
    normalized_data = map_abi_data(normalizers, output_types, output_data)
  File "/usr/local/lib/python3.6/site-packages/eth_utils/string.py", line 85, in inner
    return force_obj_to_text(fn(*args, **kwargs))
  File "/usr/local/lib/python3.6/site-packages/web3/utils/abi.py", line 437, in map_abi_data
    return pipe(data, *pipeline)
  File "cytoolz/functoolz.pyx", line 586, in cytoolz.functoolz.pipe (cytoolz/functoolz.c:10663)
  File "cytoolz/functoolz.pyx", line 562, in cytoolz.functoolz.c_pipe (cytoolz/functoolz.c:10494)
  File "/usr/local/lib/python3.6/site-packages/web3/utils/decorators.py", line 31, in wrapped
    wrapped_val = to_wrap(*args)
  File "/usr/local/lib/python3.6/site-packages/web3/utils/formatters.py", line 117, in recursive_map
    items_mapped = map_collection(recurse, data)
  File "/usr/local/lib/python3.6/site-packages/web3/utils/formatters.py", line 104, in map_collection
    return datatype(map(func, collection))
  File "/usr/local/lib/python3.6/site-packages/web3/utils/formatters.py", line 116, in recurse
    return recursive_map(func, item)
  File "/usr/local/lib/python3.6/site-packages/web3/utils/decorators.py", line 29, in wrapped
    raise ValueError('Recursively called %s with %r' % (to_wrap, args))
ValueError: Recursively called <function recursive_map at 0x7f012ea030d0> with (<function strip_abi_type at 0x7f012ea06ae8>, ABITypedData(abi_type='address', data='0x168910909606A2Fca90D4C28Fa39b50407b9C526'))
```

and

```
  File "/usr/local/lib/python3.6/site-packages/web3/contract.py", line 870, in call_contract_function
    normalized_data = map_abi_data(normalizers, output_types, output_data)
  File "/usr/local/lib/python3.6/site-packages/eth_utils/string.py", line 85, in inner
    return force_obj_to_text(fn(*args, **kwargs))
  File "/usr/local/lib/python3.6/site-packages/web3/utils/abi.py", line 437, in map_abi_data
    return pipe(data, *pipeline)
  File "cytoolz/functoolz.pyx", line 586, in cytoolz.functoolz.pipe (cytoolz/functoolz.c:10663)
  File "cytoolz/functoolz.pyx", line 562, in cytoolz.functoolz.c_pipe (cytoolz/functoolz.c:10494)
  File "/usr/local/lib/python3.6/site-packages/web3/utils/decorators.py", line 29, in wrapped
    raise ValueError('Recursively called %s with %r' % (to_wrap, args))
ValueError: Recursively called <function recursive_map at 0x7f012ea030d0> with (<function strip_abi_type at 0x7f012ea06ae8>, [ABITypedData(abi_type='uint256', data=1000000000000000000), ABITypedData(abi_type='address', data='0xC66eA802717bFb9833400264Dd12c2bCeAa34a6d'), ABITypedData(abi_type='uint256', data=100000000000000000000), ABITypedData(abi_type='address', data='0xECF8F87f810EcF450940c9f60066b4a7a501d6A7'), ABITypedData(abi_type='address', data='0xbe69Be9133DAA77AeAFcA0d6330c7Ba44f597b15'), ABITypedData(abi_type='bool', data=True), ABITypedData(abi_type='uint64', data=1505245070)])
```

We are connecting to a Parity node.

### How can it be fixed?

My suspicion is that because `@reject_recursive_repeats` keeps state in `to_wrap.__already_called`, this is not thread-safe and will fail if two threads want to parse the same data at the same time.

**Comment by carver** [comment]
> My suspicion is that because @reject_recursive_repeats keeps state in to_wrap.__already_called, this is not thread-safe and will fail if two threads want to parse the same data at the same time.

Yup, that looks plausible.

Because this decorator is a runtime assertion that should never be violated, the assertion acts as a helpful hint for debugging issues, rather than as a critical runtime feature. (only the assertion is buggy here, not the actual recursive call).

As a short term workaround, you could just remove the assertion by removing the decorator at: https://github.com/pipermerriam/web3.py/blob/v3/web3/utils/formatters.py#L109 Of course that means futzing with the source code of your installed web3. So depending on how urgent your need is, you might just prefer to wait for the next stable release. I would expect that we can get this done and released sometime by end of week. (hopefully sooner)

**Comment by carver** [comment]
Released in v3.16.3
